Doc/howto/a-conceptual-overview-of-asyncio.rst

```diff
@@ -0,0 +1,502 @@
+We want that task to finish only after three seconds have elapsed, but without
+preventing other tasks from running.
+
+::
+
+   async def other_work():
+       print("I am worker. Work work.")
+
+   async def main():
+       # Add a few other tasks to the event loop, so there's something
+       # to do while asynchronously sleeping.
+       work_tasks = [
+           asyncio.create_task(other_work()),
+           asyncio.create_task(other_work()),
+           asyncio.create_task(other_work())
+       ]
+       print(
+           "Beginning asynchronous sleep at time: "
+           f"{datetime.datetime.now().strftime("%H:%M:%S")}."
+       )
+       await asyncio.create_task(async_sleep(3))
+       print(
+           "Done asynchronous sleep at time: "
+           f"{datetime.datetime.now().strftime("%H:%M:%S")}."
+       )
+       # asyncio.gather effectively awaits each task in the collection.
+       await asyncio.gather(*work_tasks)
+
+
+Below, we use a future to enable custom control over when that task will be
+marked as done.
+If :meth:`future.set_result() <asyncio.Future.set_result>` (the method
+responsible for marking that future as done) is never called, then this task
+will never finish.
+We've also enlisted the help of another task, which we'll see in a moment, that
+will monitor how much time has elapsed and accordingly call
+``future.set_result()``.
+
+::
+
+   async def async_sleep(seconds: float):
+       future = asyncio.Future()
+       time_to_wake = time.time() + seconds
+       # Add the watcher-task to the event loop.
+       watcher_task = asyncio.create_task(_sleep_watcher(future, time_to_wake))
+       # Block until the future is marked as done.
+       await future
+
+We'll use a rather bare object, ``YieldToEventLoop()``, to ``yield`` from
+``__await__`` in order to cede control to the event loop.
+This is effectively the same as calling ``asyncio.sleep(0)``, but this approach
+offers more clarity, not to mention it's somewhat cheating to use
+``asyncio.sleep`` when showcasing how to implement it!
+
+As usual, the event loop cycles through its queue of tasks, giving them control
+and receiving control back when they pause or finish.
+The ``watcher_task``, which runs the coroutine ``_sleep_watcher(...)``, will
+be invoked once per full cycle of the event loop's queue.
+On each resumption, it'll check the time and if not enough has elapsed, then
+it'll pause once again and hand control back to the event loop.
+Eventually, enough time will have elapsed, and ``_sleep_watcher(...)`` will
+mark the future as done, and then itself finish too by breaking out of the
+infinite ``while`` loop.
+Given this helper task is only invoked once per cycle of the event loop's queue,
+you'd be correct to note that this asynchronous sleep will sleep *at least*
+three seconds, rather than exactly three seconds.
+Note this is also of true of ``asyncio.sleep``.
+
+::
+
+   class YieldToEventLoop:
+       def __await__(self):
```

Comment:
Bare yield is used internally by asyncio to implement asyncio.sleep(0) and is special cased, I don't think this is a good example to document as it relies too much on implementation detail.

Reply:
> No, asyncio libraries are built on top of futures and tasks and not objects implementing `__await__` from scratch.

I think they mean alternatives to asyncio such as trio, not asyncio libraries. The point is to understand how asyncio actually works, rather than the bare minimum needed to build something on top of it. Understanding this point helps advanced users understand when (and why) code may or may not yield to the event loop.

Reply:
> Why? Is this documented somewhere?

> yielding None is special cased and is used for relinquishing control for one event loop iteration. You can see this in implementation of tasks.

I'm not sure! I haven't checked the docs. And yes I imagine it is! But that's not really my point. `asyncio` happens to use `yield`ing from an `__await__` to cede control. But, that's also the only way to cede control from a coroutine. This is illustrated in the example earlier in the article which features no usage of `asyncio` nor an `event loop`.